lib/kinetic_cafe/error_tasks.rb

```diff
@@ -1,102 +1,144 @@
+  end
 
-      if @descendants[StandardError]
-        @descendants[StandardError].sort_by(&:to_s).each do |d| traverse.(d) end
+  def collect_translations(descendants, translations, root)
+    translation = (translations[root.i18n_key_base] ||= {})
+    name = KineticCafe::ErrorDSL.namify(root)
 
-        # Normal YAML dump does not match the pattern that we want to compare
-        # against. Therefore, we are going to write this file manually.
-        text = 'kc:'
+    params = root.i18n_params.map { |param| "%{#{param}}" }.join(' ')
 
-        translations.keys.sort.each do |group|
-          text << "\n  #{group}:"
+    translation[name] = if params.empty?
+                          "Translation for #{name} with no params."
+                        else
+                          "Translation for #{name} with #{params}."
+                        end
 
-          translations[group].keys.sort.each { |k|
-            text << "\n    #{k}: >-\n      #{translations[group][k]}"
-          }
-        end
+    if descendants[root]
+      descendants[root].sort_by(&:to_s).each do |child|
+        collect_translations(descendants, translations, child)
+      end
+    end
+  end
 
-        if args.output
-          File.open(args.output, 'w') { |f| f.puts text }
-        else
-          puts text
-        end
+  def find_translations(descendants)
+    {}.tap do |translations|
+      descendants[::StandardError].sort_by(&:to_s).each do |d|
+        collect_translations(descendants, translations, d)
       end
     end
+  end
 
-    task :find do
-      @descendants = {}
-      ObjectSpace.each_object(Class) do |k|
-        next unless k.singleton_class < KineticCafe::ErrorDSL
+  def build_translation_yaml(descendants)
+    translations = find_translations(descendants)
+    return if translations.empty?
+
+    # Normal YAML dump does not match the pattern that we want to compare
+    # against. Therefore, we are going to write this file manually.
+    String.new('kc:').tap do |text|
+      translations.keys.sort.each do |group|
+        text << "\n  #{group}:"
 
-        (@descendants[k.superclass] ||= []) << k
+        translations[group].keys.sort.each { |k|
+          text << "\n    #{k}: >-\n      #{translations[group][k]}"
+        }
       end
     end
   end
+
+  def save_translation_yaml(filename, descendants = @descendants)
+    File.open(filename, 'w') { |f| print_translation_yaml(descendants, output: f) }
+  end
+end
+
+##
+module KineticCafe::ErrorTasks #:nodoc:
+  namespace :kcerror do
+    desc 'Show defined errors.'
+    task :defined, [ :params ] => 'kcerror:find' do |_, args|
+      print_defined(params: args.params =~ /^y/i)
+    end
+
+    desc 'Generate a sample translation key file.'
+    task :translations, [ :output ] => 'kcerror:find' do |_, args|
+      if args.output
+        save_translation_yaml(args.output)
+      else
+        print_translation_yaml
```

Comment:
are there tests for this method? there are tests for the error hierarchy stuff but nothing for the translation stuff